druntime/src/core/stdc/tgmath.d

```diff
@@ -1421,12 +443,24 @@ else
     ///
     alias core.stdc.math.lroundl       lround;
 
+version (NetBSD)
+{
+    ///
+    alias core.stdc.math.llround       llroundl;
+    ///
+    alias core.stdc.math.llroundf      llroundl;
+    ///
+    alias core.stdc.math.llroundl      llroundl;
+}
+else
+{
     ///
     alias core.stdc.math.llround       llround;
     ///
     alias core.stdc.math.llroundf      llround;
     ///
     alias core.stdc.math.llroundl      llround;
+}
```

Comment:
The condition should be inverted. The default should be the platform doesn't implement these functions.

Reply:
It is the used condition is fully corresponds to current ~master tgmath.d: NetBSD uses `llroundl`, all other platforms use `llround`

Reply:
And what about AIX? CRuntime_Newlib? It's a porter's nightmare having to unpick these wrong default assumptions.

Reply:
>  It's a porter's nightmare having to unpick

@ibuclaw ah, understood your point

Reply:
I checked: tgmath.d is identical in dmd, gcc and ldc repositories

Reply:
Is that this file are generated and cannot be touched?

Reply:
@denizzzka
1. This file is hand maintained.
2. It's through nearly a decade of porting druntime and banging head for hard-to-track corruption bugs ([for example](https://github.com/dlang/dmd/blob/84ca0934647e09a755ea3f31b346a3e1b2786322/compiler/src/dmd/root/file.d#L275-L303)), that this approach of `else {/*nothing*/}` or `else static assert(false, "unsupported");` is the only way to help maintainers add new ports.
3. I have ported D to PPC64-AIX, it's sitting on a server as I don't know whether bindings can be put on GitHub. :-)

Reply:
[Another example](https://github.com/dlang/dmd/blob/84ca0934647e09a755ea3f31b346a3e1b2786322/druntime/src/core/internal/gc/impl/conservative/gc.d#L101-L102) - someone decided that OSX has a `__fork()` function, but didn't consider MacOS 10.6 which doesn't (nor any future version that might rip it out - they are making quite aggressive changes as of late).  This means bootstrap needlessly fails for D on these targets.

Reply:
But in this PR I don't change behaviour. It was like:
```d
version(NetBSD) {
  SAME_CODE;
} else version(NetBSD) {
  SAME_CODE;
} else {
  SAME_CODE;
}
```
And only few lines was different for a whole ~500 lines of SAME_CODE (same byte to byte).

Now it like:
```d
version(all) {
  SAME_CODE; // with few tiny version() branches inside
}
```

Someone used HPPA64:
https://github.com/dlang/dmd/blob/84ca0934647e09a755ea3f31b346a3e1b2786322/druntime/src/core/sys/posix/sys/msg.d#L21

But forgot to use it here and at many other places:
https://github.com/dlang/dmd/blob/84ca0934647e09a755ea3f31b346a3e1b2786322/druntime/src/core/sys/linux/dlfcn.d#L15

I also want to reduce erros, but by reducing copy-paste :-)

Reply:
Note that druntime has code duplicated per platform by design. While it's often the same signatures, when one platform changes, the update doesn't necessarily need to be copied to other platforms. druntime used to be de-duplicated and actually more error-prone because of this reason (someone would accidentally break one platform while trying to fix another platform).

Reply:
Community was considering the variant of placing files in directories by principle:
```
druntime/src/core/stdc/tgmath.d (contains `public import : tgmath_impl` )
druntime/impl/ppc64-aix/tgmath_impl.d
druntime/impl/x86-64/tgmath_impl.d
...
```
?

This may help to force de-duplicated nature of druntime and then it would be convenient to switch platforms during build (just by build script, without all theese branches of branches), add new ones (which my interest)

Reply:
Sounds good to me, it would also help with ignoring platforms you're not interested in.

Reply:
Although the details need to be hashed out before committing to a design

Reply:
(this question was deleted previously)

Community was considering the variant of placing files in directories by principle:
```
druntime/src/core/stdc/tgmath.d (contains `public import : tgmath_impl` )
druntime/impl/netbsd/tgmath_impl.d
druntime/impl/openbsd/tgmath_impl.d
...
```
?

This may help to force de-duplicated nature of druntime and then it would be convenient to switch platforms during build (just by build script, without all theese branches of branches), add new ones (which my interest)

Reply:
> Although the details need to be hashed out before committing to a design

At the beginning I would suggest not doing anything with the architectures (CPUs) and deal only with OSes and libc/Glibc/UClibc

We need to go somewhere to discuss this

Reply:
> We need to go somewhere to discuss this

https://forum.dlang.org/post/jdfwovvqdcusjphehwzx@forum.dlang.org

Reply:
> > Although the details need to be hashed out before committing to a design
> 
> At the beginning I would suggest not doing anything with the architectures (CPUs) and deal only with OSes and libc/Glibc/UClibc
> 
> We need to go somewhere to discuss this

Walter and I gave this a try a few years ago (looked up the old PRs, a decade? 😧). From what I recall it does start falling apart once you try to split up CPU-specific stuff into separate package modules.

https://github.com/dlang/druntime/pull/1255
https://github.com/dlang/druntime/pull/732

Since then I've made dozens of PRs to isolate the OS-specific parts from the C runtime.

https://github.com/dlang/druntime/pull/3687
https://github.com/dlang/druntime/pull/3685
https://github.com/dlang/druntime/pull/3679